### PR TITLE
feat: adding slug names in urls

### LINKED ISSUE #11156: [Task] Remove default params from the client-side URL builder functions

### LINKED ISSUE #11157: [Task] Initialize application based on the page id in the URL
All APIs required to initialize an Appsmith application leverages the application ID in the existing URL structure. 
We need to remove that dependency since the new URL will not have application ID in it. Instead we would use to page id to retrieve the application ID and resume the old workflow

### PR DESCRIPTION
**Pull request by arunvjn**
## Description
This PR aims to change the structure of the current URL with a more user friendly URL.
/applications/:appId/pages/:pageId -> /applicationSlugName/pageSlugName-pageId

### Following are the changes made in this PR.
- Changed the existing URL builder functions to accommodate application slug and page slug.
- URL builder functions will no longer have default params, typescript should be able to warn about invalid usage.
- Application initialisation is based on the Page ID in the URL.
- A new merged API which fetches both application and page list details using the pageId in the URL.
- Application with older URLs will be redirected to the new URLs based on the applicationVersion attribute (if > 1).
- Introduces a new upgrade option for users to update their existing application to let them fully understand the update and warn them about possible places it could break their app.
- A new route module which persists the application and page info thereby simplifying navigation operations and improve DX.

https://docs.google.com/spreadsheets/d/1AL4YDc892AguREWtgQa6vAHtogb5kW9SIF9YUpiqOMk/edit#gid=0

Fixes #11156 
Fixes #11157

## Type of change
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?
- Manual
- Jest
- Cypress

## Design
Manual upgrade - https://www.figma.com/file/uMVbwxXa5AUWcJKJ4sYqUb/Clean-URLs?node-id=3523%3A6484

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feat/better_urls 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.76 **(0.21)** | 37.21 **(0.18)** | 35.35 **(0.17)** | 56.02 **(0.17)**
 :sparkles: :new: | **app/client/src/RouteBuilder.ts** | **84.85** | **85.42** | **31.58** | **80.77**
 :green_circle: | app/client/src/store.ts | 86.36 **(1.36)** | 50 **(50)** | 80 **(30)** | 83.33 **(-0.88)**
 :green_circle: | app/client/src/actions/applicationActions.ts | 72.73 **(0)** | 100 **(0)** | 16.67 **(4.91)** | 64.71 **(0)**
 :red_circle: | app/client/src/actions/pageActions.tsx | 68.32 **(-2.16)** | 25 **(0)** | 26.67 **(-5.59)** | 58.57 **(-3.07)**
 :red_circle: | app/client/src/api/ApplicationApi.tsx | 51.06 **(-2.13)** | 0 **(0)** | 5.26 **(-5.27)** | 52.27 **(-2.28)**
 :green_circle: | app/client/src/api/PageApi.tsx | 65.38 **(1.38)** | 25 **(0)** | 21.05 **(4.38)** | 66 **(1.42)**
 :green_circle: | app/client/src/ce/constants/messages.ts | 76.95 **(0.03)** | 100 **(0)** | 31.68 **(0.2)** | 81.83 **(0.06)**
 :red_circle: | app/client/src/comments/CommentCard/CommentCard.tsx | 71.67 **(-0.31)** | 57.25 **(0)** | 44.19 **(0)** | 72.02 **(-0.33)**
 :green_circle: | app/client/src/components/ads/Icon.tsx | 99.43 **(0.01)** | 84.21 **(0)** | 100 **(0)** | 99.42 **(0)**
 :green_circle: | app/client/src/components/ads/index.ts | 100 **(0)** | 100 **(0)** | 5.56 **(3.71)** | 100 **(0)**
 :green_circle: | app/client/src/components/editorComponents/CloseEditor.tsx | 42.86 **(0.92)** | 25 **(0)** | 0 **(0)** | 42.86 **(0.92)**
 :red_circle: | app/client/src/components/editorComponents/Sidebar.tsx | 82.56 **(-0.59)** | 44.12 **(0)** | 62.5 **(0)** | 82.72 **(-0.61)**
 :green_circle: | app/client/src/components/editorComponents/ActionRightPane/index.tsx | 35.56 **(0.4)** | 25 **(0)** | 0 **(0)** | 37.65 **(0.44)**
 :red_circle: | app/client/src/components/editorComponents/Debugger/EntityLink.tsx | 26.47 **(-0.67)** | 7.41 **(0)** | 7.14 **(0)** | 28.13 **(0.19)**
 :green_circle: | app/client/src/components/editorComponents/Debugger/hooks/debuggerHooks.ts | 36.78 **(2.26)** | 0 **(0)** | 7.14 **(0)** | 35.14 **(0.89)**
 :green_circle: | app/client/src/components/editorComponents/GlobalSearch/index.tsx | 52.52 **(0.17)** | 16.84 **(0)** | 34.48 **(0)** | 54.98 **(0.18)**
 :red_circle: | app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx | 28.89 **(0)** | 2.92 **(-0.09)** | 3.23 **(0)** | 30.72 **(0)**
 :red_circle: | app/client/src/constants/AppConstants.ts | 60.71 **(-32.15)** | 0 **(-50)** | 0 **(-100)** | 60 **(-36)**
 :green_circle: | app/client/src/constants/routes.ts | 93.41 **(27.98)** | 66.67 **(42.92)** | 57.14 **(35.71)** | 94.19 **(32.1)**
 :red_circle: | app/client/src/pages/AppViewer/viewer/GetAppViewerHeaderCTA.tsx | 87.88 **(-5.45)** | 83.33 **(0)** | 60 **(-15)** | 87.88 **(-5.45)**
 :red_circle: | app/client/src/pages/AppViewer/viewer/PageTabs.tsx | 32.73 **(1.25)** | 21.43 **(-3.57)** | 0 **(0)** | 35.29 **(1.29)**
 :green_circle: | app/client/src/pages/Editor/EditorHeader.tsx | 86.9 **(0.38)** | 53.09 **(0)** | 62.5 **(0)** | 86.23 **(0.2)**
 :red_circle: | app/client/src/pages/Editor/WidgetSidebar.tsx | 70.73 **(2.44)** | 14.29 **(-7.14)** | 44.44 **(0)** | 69.23 **(2.56)**
 :red_circle: | app/client/src/pages/Editor/routes.tsx | 43.75 **(-3.39)** | 28.57 **(6.35)** | 0 **(-6.67)** | 48.28 **(-1.72)**
 :red_circle: | app/client/src/pages/Editor/APIEditor/ApiRightPane.tsx | 43.28 **(-1.01)** | 44 **(0)** | 0 **(0)** | 46.67 **(-0.95)**
 :green_circle: | app/client/src/pages/Editor/APIEditor/ProviderTemplates.tsx | 37.25 **(0.98)** | 40.82 **(0)** | 4.76 **(0)** | 38.95 **(1.06)**
 :sparkles: :new: | **app/client/src/pages/Editor/BottomBar/ManualUpgrades.tsx** | **89.36** | **81.25** | **72.73** | **87.5**
 :green_circle: | app/client/src/pages/Editor/DataSourceEditor/BackButton.tsx | 62.5 **(2.5)** | 50 **(0)** | 0 **(0)** | 62.5 **(2.5)**
 :red_circle: | app/client/src/pages/Editor/DataSourceEditor/index.tsx | 30.21 **(-0.97)** | 0 **(0)** | 18.18 **(0)** | 30.34 **(-0.34)**
 :green_circle: | app/client/src/pages/Editor/EditorAppName/NavigationMenuData.ts | 67.19 **(0.52)** | 41.67 **(0)** | 6.25 **(0)** | 66.67 **(0.54)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Datasources.tsx | 95.35 **(0.11)** | 85.71 **(0)** | 71.43 **(0)** | 95 **(0.13)**
 :red_circle: | app/client/src/pages/Editor/Explorer/EntityExplorer.tsx | 88.24 **(-0.65)** | 66.67 **(0)** | 50 **(0)** | 87.76 **(-0.7)**
 :red_circle: | app/client/src/pages/Editor/Explorer/helpers.tsx | 76.81 **(-11.87)** | 40.43 **(-19.33)** | 85.71 **(2.38)** | 75.44 **(-11.52)**
 :green_circle: | app/client/src/pages/Editor/Explorer/hooks.ts | 30.36 **(0.56)** | 26.49 **(3.04)** | 21.82 **(0)** | 30.85 **(0.7)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/ActionEntity.tsx | 45.45 **(-2.94)** | 0 **(0)** | 0 **(0)** | 51.72 **(0)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx | 36.54 **(-0.96)** | 0 **(0)** | 0 **(0)** | 38 **(-0.89)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/helpers.tsx | 35.14 **(-1.45)** | 0 **(0)** | 0 **(0)** | 35.48 **(-1.66)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Datasources/DatasourceEntity.tsx | 75 **(-1.92)** | 30 **(0)** | 50 **(0)** | 75 **(-1.92)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Datasources/QueryTemplates.tsx | 47.06 **(1.61)** | 40 **(0)** | 0 **(0)** | 48.48 **(1.6)**
 :green_circle: | app/client/src/pages/Editor/Explorer/JSActions/JSActionEntity.tsx | 60.87 **(2.54)** | 0 **(0)** | 0 **(0)** | 60.87 **(2.54)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Pages/index.tsx | 83.05 **(1.38)** | 75 **(8.33)** | 42.86 **(0)** | 83.05 **(1.38)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Widgets/useNavigateToWidget.ts | 88.89 **(-0.47)** | 70 **(0)** | 100 **(0)** | 90.7 **(-0.41)**
 :green_circle: | app/client/src/pages/Editor/FirstTimeUserOnboarding/Checklist.tsx | 96.27 **(0.03)** | 96.61 **(0)** | 92.68 **(0)** | 96.27 **(0.03)**
 :red_circle: | app/client/src/pages/Editor/FirstTimeUserOnboarding/Statusbar.tsx | 89.89 **(-0.22)** | 77.46 **(0)** | 85.71 **(0)** | 89.66 **(-0.23)**
 :green_circle: | app/client/src/pages/Editor/FirstTimeUserOnboarding/Tasks.tsx | 98.46 **(0.02)** | 96.43 **(0)** | 91.67 **(0)** | 98.41 **(0.02)**
 :red_circle: | app/client/src/pages/Editor/GeneratePage/components/ActionCards.tsx | 42.42 **(-5.86)** | 100 **(0)** | 0 **(0)** | 46.67 **(-7.18)**
 :green_circle: | app/client/src/pages/Editor/GeneratePage/components/GeneratePageForm/GeneratePageForm.tsx | 18.78 **(0.36)** | 10.53 **(0)** | 0 **(0)** | 20.98 **(0.39)**
 :red_circle: | app/client/src/pages/Editor/IntegrationEditor/DatasourceCard.tsx | 73.49 **(-1.51)** | 68.57 **(0)** | 14.29 **(0)** | 72.84 **(-1.58)**
 :red_circle: | app/client/src/pages/Editor/IntegrationEditor/IntegrationsHomeScreen.tsx | 27.22 **(-0.59)** | 11.36 **(0)** | 4.55 **(0)** | 29.37 **(-3.17)**
 :red_circle: | app/client/src/pages/Editor/IntegrationEditor/NewApi.tsx | 34.21 **(-0.93)** | 20 **(0)** | 0 **(0)** | 37.14 **(0)**
 :red_circle: | app/client/src/pages/Editor/JSEditor/index.tsx | 54.84 **(-1.83)** | 33.33 **(0)** | 33.33 **(0)** | 59.26 **(-2.28)**
 :green_circle: | app/client/src/pages/Editor/PagesEditor/index.tsx | 47.83 **(1.02)** | 60 **(0)** | 0 **(0)** | 48.89 **(1.06)**
 :red_circle: | app/client/src/pages/Editor/PropertyPane/ConnectDataCTA.tsx | 82.86 **(-0.92)** | 100 **(0)** | 85.71 **(0)** | 82.35 **(-0.98)**
 :green_circle: | app/client/src/pages/Editor/QueryEditor/index.tsx | 31.48 **(0.92)** | 7.84 **(0)** | 5.88 **(0)** | 32.67 **(0.99)**
 :green_circle: | app/client/src/pages/Editor/SaaSEditor/ActionForm.tsx | 26.76 **(0.37)** | 0 **(0)** | 0 **(0)** | 29.23 **(0.44)**
 :green_circle: | app/client/src/pages/Editor/SaaSEditor/DatasourceCard.tsx | 61.11 **(0.58)** | 52.94 **(0)** | 0 **(0)** | 61.11 **(0.58)**
 :green_circle: | app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx | 38.71 **(0.61)** | 3.92 **(0)** | 14.29 **(0)** | 41.07 **(0.72)**
 :red_circle: | app/client/src/pages/Editor/SaaSEditor/ListView.tsx | 34.57 **(-0.49)** | 32 **(0)** | 6.25 **(0)** | 38.03 **(0.35)**
 :green_circle: | app/client/src/pages/Editor/SaaSEditor/constants.ts | 100 **(54.29)** | 100 **(100)** | 100 **(100)** | 100 **(61.29)**
 :sparkles: :new: | **app/client/src/pages/tests/mockData.ts** | **100** | **100** | **100** | **100**
 :red_circle: | app/client/src/reducers/entityReducers/appReducer.ts | 54.55 **(-9.09)** | 100 **(0)** | 16.67 **(-16.66)** | 54.55 **(-9.09)**
 :green_circle: | app/client/src/reducers/uiReducers/applicationsReducer.tsx | 15 **(5)** | 27.27 **(0)** | 10.2 **(2.04)** | 13.4 **(3.09)**
 :green_circle: | app/client/src/sagas/ActionSagas.ts | 12.57 **(0.24)** | 1.5 **(0)** | 8 **(0)** | 15.33 **(0.15)**
 :green_circle: | app/client/src/sagas/ApiPaneSagas.ts | 17.62 **(1.03)** | 1.19 **(0)** | 9.09 **(0)** | 19.41 **(0.66)**
 :green_circle: | app/client/src/sagas/ApplicationSagas.tsx | 20.81 **(1.76)** | 4.38 **(-1.56)** | 9.09 **(2.19)** | 23.89 **(1.77)**
 :green_circle: | app/client/src/sagas/CurlImportSagas.ts | 50 **(1.43)** | 33.33 **(0)** | 50 **(0)** | 55.17 **(0.33)**
 :green_circle: | app/client/src/sagas/DatasourcesSagas.ts | 13.98 **(0.65)** | 2.03 **(0)** | 8.33 **(0)** | 16.42 **(0.46)**
 :red_circle: | app/client/src/sagas/EvaluationsSaga.ts | 19.3 **(-0.88)** | 1.94 **(0)** | 7.41 **(-3.7)** | 22 **(-1)**
 :green_circle: | app/client/src/sagas/GitSyncSagas.ts | 10.16 **(0.29)** | 0.26 **(0)** | 4.35 **(0)** | 12.75 **(0.35)**
 :green_circle: | app/client/src/sagas/InitSagas.ts | 48.65 **(8.53)** | 29.51 **(13.99)** | 58.33 **(-8.34)** | 52.6 **(8.54)**
 :green_circle: | app/client/src/sagas/JSActionSagas.ts | 17.83 **(0.44)** | 2.13 **(0)** | 6.67 **(0)** | 20.93 **(0.32)**
 :green_circle: | app/client/src/sagas/JSPaneSagas.ts | 15.42 **(0.56)** | 1.11 **(0)** | 4.76 **(0)** | 17.93 **(0.38)**
 :red_circle: | app/client/src/sagas/OnboardingSagas.ts | 21.05 **(0)** | 1.06 **(-0.1)** | 4.35 **(0)** | 23.88 **(0)**
 :green_circle: | app/client/src/sagas/PageSagas.tsx | 24.14 **(0.21)** | 5.38 **(0)** | 14.29 **(0.4)** | 25.67 **(0.09)**
 :green_circle: | app/client/src/sagas/PluginSagas.ts | 14 **(0.67)** | 2.17 **(0)** | 8.33 **(-0.76)** | 16.47 **(0.19)**
 :green_circle: | app/client/src/sagas/QueryPaneSagas.ts | 25.98 **(2.79)** | 8.33 **(0)** | 18.18 **(0)** | 29.63 **(2.2)**
 :red_circle: | app/client/src/sagas/RecentEntitiesSagas.ts | 56.41 **(-19.27)** | 12.2 **(-2.69)** | 75 **(0)** | 54.05 **(-20.95)**
 :green_circle: | app/client/src/sagas/SaaSPaneSagas.ts | 43.48 **(5.3)** | 25 **(0)** | 50 **(0)** | 50 **(3.49)**
 :green_circle: | app/client/src/sagas/SnipingModeSagas.ts | 19.81 **(0.37)** | 3.03 **(0)** | 40 **(0)** | 20.39 **(0.2)**
 :red_circle: | app/client/src/sagas/TemplatesSagas.ts | 26.67 **(-0.16)** | 5.56 **(-2.77)** | 14.29 **(-2.38)** | 32.43 **(0.08)**
 :green_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 51.76 **(0.22)** | 38.73 **(0)** | 52.08 **(0)** | 52.93 **(0.14)**
 :green_circle: | app/client/src/sagas/selectors.tsx | 78 **(1)** | 58.33 **(0)** | 62.5 **(3.12)** | 73.68 **(1.31)**
 :green_circle: | app/client/src/sagas/ActionExecution/NavigateActionSaga.ts | 36.36 **(1.58)** | 8 **(0)** | 25 **(0)** | 39.02 **(0.92)**
 :green_circle: | app/client/src/sagas/ActionExecution/PluginActionSaga.ts | 18.25 **(0.23)** | 2.31 **(0)** | 9.52 **(0)** | 20.99 **(0.24)**
 :green_circle: | app/client/src/selectors/appViewSelectors.tsx | 69.23 **(0.48)** | 0 **(0)** | 33.33 **(4.76)** | 66.67 **(0)**
 :red_circle: | app/client/src/selectors/commentsSelectors.ts | 83.61 **(-1.64)** | 61.76 **(-2.95)** | 73.33 **(0)** | 88.24 **(-2.35)**
 :red_circle: | app/client/src/selectors/editorSelectors.tsx | 75.32 **(-0.97)** | 61.74 **(-8.39)** | 60 **(-1.54)** | 73.66 **(-0.39)**
 :red_circle: | app/client/src/utils/AppsmithUtils.tsx | 54.17 **(0.2)** | 28.33 **(-2.03)** | 35.42 **(0)** | 49.51 **(0.24)**
 :red_circle: | app/client/src/utils/DSLMigrations.ts | 73.14 **(-0.24)** | 69.27 **(0)** | 71.13 **(0)** | 73.05 **(-0.14)**
 :green_circle: | app/client/src/utils/helpers.tsx | 66.98 **(2.89)** | 34.36 **(5.08)** | 50.88 **(2.73)** | 64.8 **(2.9)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :green_circle: | app/client/src/utils/hooks/localstorage.tsx | 63.16 **(42.11)** | 25 **(25)** | 66.67 **(66.67)** | 58.82 **(35.29)**</details>

### CODE REVIEW

**Review by akash-codemonk — CHANGES_REQUESTED** [review]

**Comment by arunvjn** [comment]
@mohanarpit Could you please review the backend changes on this PR ?

File: app/client/src/pages/Editor/BottomBar/ManualUpgrades.tsx [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+          ))}
+        </ul>
+      </div>
+    ) : (
+      "No new updates"
+    );
+
+  return (
+    <div className="t--upgrade relative">
+      {applicationVersion < latestVersion && <RedDot />}
```
[the remark this answers is not in the document]
Comment (arunvjn) [reply app/client/src/pages/Editor/BottomBar/ManualUpgrades.tsx:83]: DP points to release, update works with the backend changes on this branch.

File: app/client/src/pages/Editor/BottomBar/ManualUpgrades.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  selectApplicationVersion,
+} from "selectors/editorSelectors";
+import styled from "styled-components";
+import { useLocalStorage } from "utils/hooks/localstorage";
+
+const updates = [
+  {
+    name: "Update URLs",
+    shortDesc:
+      "All URLs in your applications will update to a new readable format that includes the application and page names.",
```
[the remark this answers is not in the document]
Comment (arunvjn) [reply app/client/src/pages/Editor/BottomBar/ManualUpgrades.tsx]: Done

File: app/client/src/sagas/ApplicationSagas.tsx [hunk 3]
```diff
@@ -637,8 +695,10 @@ export function* importApplicationSaga(
           );
         } else {
           const defaultPage = pages.filter((eachPage) => !!eachPage.isDefault);
-          const pageURL = BUILDER_PAGE_URL({
-            applicationId: id,
+          const pageURL = builderURL({
+            //Comeback
```
[the remark this answers is not in the document]
Comment (arunvjn) [reply app/client/src/sagas/ApplicationSagas.tsx]: Done

**Review by akash-codemonk — DISMISSED** [review]

**Review by akash-codemonk — DISMISSED** [review]

**Review by akash-codemonk — DISMISSED** [review]

**Review by sharat87 — APPROVED** [review]

**Review by akash-codemonk — APPROVED** [review]

**Pull request merged**